version: '3.9'

services:

  mongodb:
    image: mongo:latest
    ports:
      - "27017:27017"
    networks:
      - usernetwork
    restart: always
    environment:
      MONGO_INITDB_DATABASE: datalogs
    healthcheck:
      test: ["CMD", "mongo", "--eval", "db.runCommand({ ping: 1 })"]
      interval: 5s
      retries: 5
      timeout: 10s

  mysqldb:
    image: mysql:8.0
    environment:
      MYSQL_ROOT_PASSWORD: 1234567
      MYSQL_DATABASE: desafio_3
      TZ: UTC
      MYSQL_USER: docker
      MYSQL_PASSWORD: 1234567
    healthcheck:
      test: ["CMD", "mysqladmin", "-u$MYSQL_USER", "-p$MYSQL_PASSWORD",  "ping", "-h", "localhost"]
      interval: 5s
      retries: 5
      timeout: 10s

    ports:
      - "3306:3306"
    networks:
      - usernetwork
    restart: always

  rabbitmq:
    image: rabbitmq:3.9-management
    ports:
      - "5672:5672"
      - "15672:15672"
    networks:
      - usernetwork
    healthcheck:
      test: ["CMD", "rabbitmqctl", "status"]
      interval: 5s
      retries: 5
      timeout: 10s

  user-management:
    image: user-management
    build: ./thirdchallenge
    working_dir: /thirdchallenge
    ports:
      - "8080:8080"
    depends_on:
      mysqldb:
        condition: service_started
      rabbitmq:
        condition: service_healthy
    restart: always
    networks:
      - usernetwork
    environment:
      - SPRING_DATASOURCE_URL=jdbc:mysql://mysqldb:3306/desafio_3
      - SPRING_DATASOURCE_USERNAME=root
      - SPRING_DATASOURCE_PASSWORD=1234567
      - SECRET_KEY=70165810360642928052074927762786

  notify:
    image: notify
    working_dir: /notify
    build: ./notify
    depends_on:
      rabbitmq:
        condition: service_healthy
      mysqldb:
        condition: service_started
    networks:
      - usernetwork

networks:
  usernetwork:
    driver: bridge